### PR TITLE
Make page require faster

### LINKED ISSUE #3682: Performance issue with /server/resolve.js
- [x] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Expected Behavior

The server shouldn't spend 20ms to require the `Component` and the `Document` components.
I believe we could **use a cache to only pay the resolution cost once in production.**

## Current Behavior

The requests are delayed by 20ms.
It might get worse in the future version as the [`/server/resolve.js`](https://github.com/zeit/next.js/blob/cdbbe928b26af2e79c4508f127351b6652c5aee6/server/resolve.js) logic has grown to support 5 different extensions in canary:
https://github.com/zeit/next.js/blob/cdbbe928b26af2e79c4508f127351b6652c5aee6/server/resolve.js#L36-L42

where it only supports 2 in 4.2.3
https://github.com/zeit/next.js/blob/44de0f15c96c425c193ffdccacb14f28051051d5/server/resolve.js#L33-L36

## Steps to Reproduce (for bugs)

1. Server side render a page with the `--inspect` flag

## Context

I have noticed the following issue investigating a server-side rendering slow issue.
I have noticed that we systematically wait 20ms to require the `Component` and the `Document` components.

https://github.com/zeit/next.js/blob/44de0f15c96c425c193ffdccacb14f28051051d5/server/render.js#L56-L59

I believe a simple cache in production could make the resolution instant:

capture d ecran 2018-02-05 a 14 43 53

## Your Environment

| Tech    | Version |
|---------|---------|
| next    | 4.2.3 |
| node    | 9.4.0 |
| OS      | macOS High Sierra 10.13.x |
| browser | N.A         |
| etc     |         |

### PR DESCRIPTION
**Pull request by timneutkens**
Fixes #3682

Results:

```
t: 0.197ms
t: 0.041ms
t: 0.030ms
t: 0.023ms
t: 0.032ms
t: 0.026ms
t: 0.030ms
t: 0.026ms
t: 0.047ms
t: 0.025ms
t: 0.025ms
t: 0.023ms
t: 0.121ms
t: 0.022ms
t: 0.025ms
t: 0.032ms
t: 0.027ms
t: 0.022ms
t: 0.070ms
t: 0.026ms
t: 0.044ms
t: 0.034ms
t: 0.037ms
t: 0.033ms
t: 0.030ms
t: 0.032ms
t: 1.030ms
t: 0.022ms
t: 0.025ms
t: 0.020ms
t: 0.024ms
t: 0.029ms
t: 0.025ms
t: 0.020ms
t: 0.026ms
t: 0.033ms
t: 0.046ms
t: 0.055ms
t: 0.022ms
t: 0.017ms
t: 0.025ms
t: 0.025ms
t: 0.024ms
t: 0.024ms
t: 0.023ms
t: 0.019ms
t: 0.401ms
t: 0.021ms
t: 0.025ms
t: 0.017ms
t: 0.023ms
t: 0.074ms
t: 0.026ms
t: 0.021ms
t: 0.024ms
t: 0.021ms
t: 0.026ms
t: 0.021ms
t: 0.020ms
t: 0.026ms
t: 0.025ms
t: 0.021ms
t: 0.022ms
t: 0.019ms
t: 0.025ms
t: 0.023ms
t: 0.024ms
t: 0.022ms
t: 0.023ms
t: 0.023ms
t: 0.024ms
t: 0.024ms
t: 0.022ms
t: 0.028ms
t: 0.022ms
t: 0.024ms
t: 0.024ms
t: 0.022ms
t: 0.028ms
t: 0.023ms
t: 0.025ms
t: 0.021ms
t: 0.024ms
t: 0.020ms
t: 0.025ms
t: 0.025ms
t: 0.120ms
t: 0.053ms
t: 0.112ms
t: 0.023ms
t: 0.025ms
t: 0.024ms
t: 0.022ms
t: 0.019ms
t: 0.026ms
t: 0.845ms
t: 0.026ms
t: 0.021ms
t: 0.042ms
t: 0.033ms
t: 0.026ms
t: 0.019ms
t: 0.024ms
t: 0.023ms
t: 0.026ms
t: 0.020ms
t: 0.024ms
t: 0.023ms
t: 0.021ms
t: 0.020ms
t: 0.023ms
t: 0.022ms
t: 0.053ms
t: 0.025ms
t: 0.029ms
t: 0.018ms
t: 0.023ms
t: 0.021ms
t: 0.057ms
t: 0.030ms
t: 0.022ms
t: 0.019ms
t: 0.052ms
t: 0.022ms
t: 0.023ms
t: 0.017ms
t: 0.025ms
t: 0.023ms
t: 0.023ms
t: 0.021ms
t: 0.025ms
t: 0.096ms
t: 0.026ms
t: 0.019ms
t: 0.064ms
t: 0.025ms
t: 0.026ms
t: 0.020ms
t: 0.027ms
t: 0.030ms
t: 0.024ms
t: 0.015ms
t: 0.026ms
t: 0.025ms
t: 0.030ms
t: 0.029ms
t: 0.030ms
t: 0.021ms
t: 0.034ms
t: 0.024ms
t: 0.028ms
t: 0.025ms
t: 0.601ms
t: 0.024ms
t: 0.029ms
t: 0.026ms
t: 0.027ms
t: 0.025ms
t: 0.024ms
t: 0.112ms
t: 0.028ms
t: 0.022ms
t: 0.027ms
t: 0.023ms
t: 0.026ms
t: 0.021ms
t: 0.038ms
t: 0.030ms
t: 0.018ms
t: 0.021ms
t: 0.026ms
t: 0.025ms
t: 0.028ms
t: 0.043ms
t: 0.024ms
t: 0.023ms
t: 0.028ms
t: 0.024ms
t: 0.054ms
t: 0.024ms
t: 2.162ms
t: 0.261ms
t: 0.024ms
t: 0.017ms
t: 0.025ms
t: 0.029ms
t: 0.025ms
t: 0.031ms
t: 0.029ms
t: 0.042ms
t: 0.026ms
t: 0.024ms
t: 0.027ms
t: 0.193ms
t: 0.023ms
t: 0.022ms
t: 0.026ms
t: 0.027ms
t: 0.321ms
t: 0.023ms
```

Old: 

```
component fetching: 27.335ms
component fetching: 11.558ms
component fetching: 2.292ms
component fetching: 17.421ms
component fetching: 2.629ms
component fetching: 5.179ms
component fetching: 3.961ms
component fetching: 3.998ms
component fetching: 4.922ms
component fetching: 4.229ms
component fetching: 1.790ms
component fetching: 3.081ms
component fetching: 2.728ms
component fetching: 3.907ms
component fetching: 13.300ms
component fetching: 2.233ms
component fetching: 8.058ms
component fetching: 1.657ms
component fetching: 2.269ms
component fetching: 2.376ms
component fetching: 1.650ms
component fetching: 1.671ms
component fetching: 4.337ms
component fetching: 5.558ms
component fetching: 1.759ms
component fetching: 1.480ms
component fetching: 1.976ms
component fetching: 2.299ms
component fetching: 5.189ms
component fetching: 4.323ms
component fetching: 1.980ms
component fetching: 1.661ms
component fetching: 22.161ms
component fetching: 7.424ms
component fetching: 7.250ms
component fetching: 1.503ms
component fetching: 1.699ms
component fetching: 6.749ms
component fetching: 1.990ms
component fetching: 3.022ms
component fetching: 6.880ms
component fetching: 3.279ms
component fetching: 1.786ms
component fetching: 2.764ms
component fetching: 1.579ms
component fetching: 1.482ms
component fetching: 2.081ms
component fetching: 3.501ms
component fetching: 1.551ms
component fetching: 1.743ms
component fetching: 2.351ms
component fetching: 6.141ms
component fetching: 1.977ms
component fetching: 1.456ms
component fetching: 1.553ms
component fetching: 3.221ms
component fetching: 1.388ms
component fetching: 2.732ms
component fetching: 2.252ms
component fetching: 1.563ms
component fetching: 1.801ms
component fetching: 4.789ms
component fetching: 1.456ms
component fetching: 1.754ms
component fetching: 1.224ms
component fetching: 1.647ms
component fetching: 1.162ms
component fetching: 4.631ms
component fetching: 1.256ms
component fetching: 1.483ms
component fetching: 1.462ms
component fetching: 1.647ms
component fetching: 1.690ms
component fetching: 1.950ms
component fetching: 5.786ms
component fetching: 1.671ms
component fetching: 6.056ms
component fetching: 1.947ms
component fetching: 1.882ms
component fetching: 3.140ms
component fetching: 1.996ms
component fetching: 2.885ms
component fetching: 2.292ms
component fetching: 4.953ms
component fetching: 3.683ms
component fetching: 1.256ms
component fetching: 1.469ms
component fetching: 1.808ms
component fetching: 1.986ms
component fetching: 2.724ms
component fetching: 1.756ms
component fetching: 2.551ms
component fetching: 3.197ms
component fetching: 2.969ms
component fetching: 2.357ms
component fetching: 3.384ms
component fetching: 3.233ms
component fetching: 1.899ms
component fetching: 2.082ms
component fetching: 2.347ms
```

### CODE REVIEW

**Review by arunoda — APPROVED** [review]
LGTM

**Comment by arunoda** [comment]
I'll check why this is failing on Windows?

**Comment by timneutkens** [comment]
@arunoda please do 🙏

**Comment by arunoda** [comment]
@timneutkens the `resolve` logic you used doesn't work well inside Windows as you expected.
See:

resolve

May be you can check whether the page stays inside the .next directory. I think with that, we don't need these kind of extra logic.
(We may already have working logic for that)

File: server/require.js [hunk 1]
```diff
@@ -1,6 +1,61 @@
...
+  return page
+}
+
+export function getPagePath (page, {dir, dist}) {
+  const pageBundlesPath = join(dir, dist, 'dist', 'bundles', 'pages')
+
+  try {
+    page = normalizePagePath(page)
+  } catch (err) {
+    throw pageNotFoundError(page)
```
Comment (arunoda) [inline server/require.js]: we need to print this `err` object.
Otherwise it's hard to debug the issue.

File: server/require.js [hunk 2]
```diff
@@ -1,6 +1,61 @@
...
+  }
+
+  const pagePath = join(pageBundlesPath, page) // Path to the page that is to be loaded
+
+  // Don't allow wandering outside of the bundles directory
+  const pathDir = parse(pagePath).dir
+  if (pathDir.indexOf(pageBundlesPath) !== 0) {
+    throw pageNotFoundError(page)
```
Comment (arunoda) [inline server/require.js]: Shall we do a `console.warn` here showing what's going on here. So, it's easy to debug.

File: server/require.js [hunk 3]
```diff
@@ -1,6 +1,61 @@
...
+
+  return pagePath
+}
+
+export default function requirePage (page, {dir, dist}) {
+  const pagePath = getPagePath(page, {dir, dist})
+  try {
+    return require(pagePath)
+  } catch (err) {
+    throw pageNotFoundError(page)
```
Comment (arunoda) [inline server/require.js]: Again, let's log the error.

**Review by arunoda — DISMISSED** [review]
@timneutkens could you do these simple error handling improvements?

**Comment by timneutkens** [comment]
Added the console.error for those throws.

**Comment by oliviertassinari** [comment]
@timneutkens Man, thank you!

**Comment by timneutkens** [comment]
Logic was improved here: https://github.com/zeit/next.js/pull/3976

**Pull request merged**